enclave/crypto/openssl/asn1.c

```diff
@@ -0,0 +1,270 @@
+        memset(sequence, 0, sizeof(oe_asn1_t));
+
+    if (!oe_asn1_is_valid(asn1) || !sequence)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    OE_CHECK(oe_asn1_get_raw(asn1, &tag, &data, &length));
+
+    if (tag != OE_ASN1_TAG_SEQUENCE)
+        OE_RAISE(OE_FAILURE);
+
+    oe_asn1_init(sequence, data, length);
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+oe_result_t oe_asn1_get_integer(oe_asn1_t* asn1, int* value)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    oe_asn1_tag_t tag;
+    const uint8_t* data;
+    size_t length;
+
+    if (value)
+        *value = 0;
+
+    if (!oe_asn1_is_valid(asn1) || !value)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    OE_CHECK(oe_asn1_get_raw(asn1, &tag, &data, &length));
+
+    if (tag != OE_ASN1_TAG_INTEGER)
+        OE_RAISE(OE_FAILURE);
+
+    /* Extract the varying-length integer one byte at a time. */
+    while (length--)
+        *value = (*value << 8) | *data++;
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+oe_result_t oe_asn1_get_oid(oe_asn1_t* asn1, oe_oid_string_t* oid)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    ASN1_OBJECT* obj = NULL;
+
+    if (oid)
+        memset(oid, 0, sizeof(oe_oid_string_t));
+
+    if (!oe_asn1_is_valid(asn1) || !oid)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Get the OID and covert it to string */
+    {
+        const unsigned char* ptr = asn1->ptr;
+
+        /* Convert OID to an ASN1 object */
+        if (!(obj = d2i_ASN1_OBJECT(&obj, &ptr, (long)oe_asn1_remaining(asn1))))
+            OE_RAISE(OE_CRYPTO_ERROR);
+
+        /* Convert OID to string format */
+        if (!OBJ_obj2txt(oid->buf, sizeof(oe_oid_string_t), obj, 1))
+            OE_RAISE(OE_CRYPTO_ERROR);
+
+        asn1->ptr = ptr;
+    }
+
+    result = OE_OK;
+
+done:
+    if (obj)
+        ASN1_OBJECT_free(obj);
+
+    return result;
+}
+
+oe_result_t oe_asn1_get_octet_string(
+    oe_asn1_t* asn1,
+    const uint8_t** data,
+    size_t* length)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    oe_asn1_tag_t tag;
+
+    if (data)
+        *data = NULL;
+
+    if (length)
+        *length = 0;
+
+    if (!oe_asn1_is_valid(asn1) || !data || !length)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    OE_CHECK(oe_asn1_get_raw(asn1, &tag, data, length));
+
+    if (tag != OE_ASN1_TAG_OCTET_STRING)
+        OE_RAISE(OE_FAILURE);
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+oe_result_t oe_asn1_string_to_date(const char* str, oe_datetime_t* date)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    char month[4];
+
+    memset(date, 0, sizeof(oe_datetime_t));
+
+    /* Convert the string to oe_datetime_t struct */
+    if (sscanf(
+            str,
+            "%3s %02u %02u:%02u:%02u %04u",
+            month,
+            &date->day,
+            &date->hours,
+            &date->minutes,
+            &date->seconds,
+            &date->year) != 6)
+    {
+        OE_RAISE(OE_FAILURE);
+    }
+
+    /* Convert the month string to integer */
```

Comment:
libc already has an implementation of this, can we just call strptime?  Do we really need to have two implementations of this?